core: add thread flags

kaspar030 opened the pull request:
This PR adds functionality to set multiple (32) "flags" per thread, and also wait for them in any combination, efficiently.

While adding 8 bytes to each TCB, this method allows faster conditional sleep than mutexes and is way more flexible. It is possible to implement mutex_wait and msg_wait with these flags, making it easy to wait for, e.g., a button to be pressed, a mutex to become free and a timer to be shot, at the same time.

This is WIP, but as people are asking (#4102), I decided to share.

OlegHahm commented:
> While adding 8 bytes to each TCB, this method allows faster conditional sleep than mutexes and is way more flexible. 

Wow, do you remember us celebrating when we could minimize tcb size by a couple of bytes a year ago or so?

I have to admit that I haven't understood the whole idea of this PR, but trading more RAM against improved runtime seems a to be a bad compromise on IoT devices.

zhuoshuguo commented:
@kaspar030 Ping, a kind reminder for rebase.

kaspar030 commented:
- rebased

kaspar030 commented:
> I have to admit that I haven't understood the whole idea of this PR

A common use case is what we're usually using mutexes for: wait for an event that might happen in an isr, eg., arrival of a network packet or a byte on a serial line.

Currently, a thread usually waits in a loop:

```
mutex_lock();
while (1) {
   mutex_lock());
   /* handle event */
 mutex_unlock();
}
```

and an ISR would unlock the mutex whenever an event occurs. You can find that kind of logic (sometimes using messages) all over the place.

While it kinda works, it has several disadvantages:

1. it is only possible to wait for one mutex, or a message, ...
2. if a message is used to pass events, and if the message queue is full, the event is lost (remember "possible lost interrupt"?)

The main concept behind this PR is that events can be atomically flagged and waited for in any combination:

```
while(1) {
  flag = thread_flag_wait_any(); /* waits for flag */
  switch(flag) {
   case FLAG_MSG:
     msg_receive(&msg);
   case FLAG_TAPDEV_ISR:
   handle_isr();
 }
}
```
  
As is, no flags for messages or mutexes are generated, but that's easy to add.

> Wow, do you remember us celebrating when we could minimize tcb size by a couple of bytes a year ago or so?

Of course! ;) I will try to move the needed struct out of tcb_t. I had that in an early version, but it was less efficient...

Kijewski commented on core/include/atomic.h:
```diff
@@ -132,6 +132,44 @@ static inline int atomic_set_to_zero(atomic_int_t *var)
+ *
+ * @return The value of *var* before the operation.
+ */
+static inline int atomic_or(atomic_int_t *var, int val)
```
Maybe make this file a PR on its own, and also add `xor` and `not` to complete the family?

OlegHahm commented:
Thanks for the explanation.

kaspar030 commented:
I've changed the implementation to use tcb->wait_data for the wait mask, and also changed the flags type to uint16_t. That way, on 32bit platforms, sizeof(tcb_t) doesn't get bigger due to alignment luck, and gains 2 byte on other platforms.

Also, I've changed core/msg.c to set some flags and xtimer_msg_receive_timeout() to make use of them.

OlegHahm commented:
Nice!

kaspar030 commented:
- changed the message flag semantics so it now always shows whether any message is waiting

@OlegHahm, would you like to review this?

Should I split the the msg changes?

OlegHahm commented on core/include/tcb.h:
```diff
@@ -38,34 +39,38 @@
     kernel_pid_t pid;           /**< thread's process id            */
     uint16_t priority;          /**< thread's priority              */
 
+    thread_flags_t flags;       /**< currently set flags            */
```
Couldn't this be optional, too, if we have a pseudomodule anyway?

kaspar030 replied:
It could. But I've also changed msg.c to use thread flags internally, which makes some things a lot nicer.

e.g., currently in xtimer_receive_msg_timeout, there's a race between receiving the message and deleting the timer, and if the thread has a message queue, that might lead to the timer message to be queued right after the actual received message.

This PR's xtimer_receive_msg_timeout() implementation uses a thread flag to pass the timeout information, solving the race problem.

The pseudomodule basically only disables the thread flags user API.

OlegHahm replied:
Ok, I see. I guess it makes sense.

OlegHahm commented on core/include/tcb.h:
```diff
@@ -38,34 +39,38 @@
-typedef struct tcb_t {
+struct tcb {
     char *sp;                   /**< thread's stack pointer         */
     uint16_t status;            /**< thread's status                */
```
Out of scope for this PR: but have ever discussed to user only 8 bits for the status?

kaspar030 replied:
Had the same thoughts... Although due to alignment, it wouldn't save much.

OlegHahm replied:
Sure - unless we find any other use for these 8 bits...

OlegHahm commented on core/include/thread_flags.h:
```diff
@@ -0,0 +1,125 @@
+thread_flags_t thread_flags_wait_one(thread_flags_t mask);
+
+/**
+ * @brief Atomically set tcb->flags | mask, used internally
```
What's the purpose of having this in the interface?

kaspar030 replied:
In an earlier version I needed that outside of thread_flags.c, so I put it in the header. Will remove.

OlegHahm commented:
> @OlegHahm, would you like to review this?

I can give it a try.

> Should I split the the msg changes?

Would make it easier to review.

kaspar030 commented:
- cut out msg and xtimer changes

kaspar030 commented on core/include/tcb.h:
```diff
@@ -1,7 +1,7 @@
  * Copyright (C) 2013 Freie Universität Berlin
  *
- * This file is subject to the terms and conditions of the GNU Lesser
+ * This file is subject to the terms and thread_conditions of the GNU Lesser
```
ups

OlegHahm commented on core/include/thread_flags.h:
```diff
@@ -0,0 +1,145 @@
+/** @} */
+
+typedef uint16_t thread_flags_t;
+typedef struct tcb tcb_t;
```
That's kinda ugly (defining it here instead of `tcb.h`), but I don't have a better idea right now, either.

OlegHahm commented on core/include/thread_flags.h:
```diff
@@ -0,0 +1,145 @@
+ * @brief Set thread flags, possibly waking it up
+ *
+ * @param[in]   tcb     thread to work on
+ * @param[in]   mask    flags to OR with thread's current flags
```
maybe `s/OR/add (logical OR)/`? Or just the `additional flags to be set for the current thread, represented as a bitmask`.

OlegHahm commented on core/include/thread_flags.h:
```diff
@@ -0,0 +1,145 @@
+/**
+ * @brief Clear current thread's flags
+ *
+ * @param[in]   mask    mask to ~& with thread's current flags
```
`unset flags for the current thread, represented as a bitmask`

(I think the current documentation is too much about the "how", and the "what" is a bit difficult to understand.)

OlegHahm commented on core/include/thread_flags.h:
```diff
@@ -0,0 +1,145 @@
+ * This API can be used to notify threads of conditions in a race-free
+ * and allocation-less way.
+ *
+ * Each thread has a number boolean flags (usually 16) in it's tcb, stored in
```
Why `usually`?

OlegHahm replied:
`s/it's/its`

OlegHahm replied:
Aren't we about to rename tcb? Maybe you could proactively use another term here, so we don't have to update the documentation (or potentially forget to do so), when we change the name.

kaspar030 replied:
```usually``` as I thought about making thread_flags_t configurable, in case someone needs some more flags...

I think keeping tcb should be fine. I promise I'll grep through the code for "tcb". ;)

OlegHahm replied:
ok

OlegHahm commented on core/include/thread_flags.h:
```diff
@@ -0,0 +1,145 @@
+ * and allocation-less way.
+ *
+ * Each thread has a number boolean flags (usually 16) in it's tcb, stored in
+ * it's flags field. Those flags can be set or unset, using thread_flags_set(),
```
`s/it's/its/`

OlegHahm commented on core/include/thread_flags.h:
```diff
@@ -0,0 +1,145 @@
+ * Those functions clear flags that caused them to return.
+ * It is not possible to wait for flags to become unset.
+ *
+ * Thread flags, apart from being very efficient, have certain properties
```
Isn't it rather "the use of thread flags" instead of "Thread flags" that makes it efficient?

kaspar030 replied:
I dropped the efficiency part. Does it sound better now?

OlegHahm commented on core/include/thread_flags.h:
```diff
@@ -0,0 +1,145 @@
+ * - setting thread flags cannot fail
+ *   If messages are used to notify a thread of a condition from within an ISR,
+ *   and the receiving thread is not waiting, has no queue or the queue is
+ *   full, the ISR cannot deliver the message. A thread flag can always be set.
```
Should we mention the downside or rather the use cases where messages must be used (e.g. if it is not only of interest **that** an event occurred, but **how many** of them occurred?

kaspar030 replied:
I've added a note pointing that out.

OlegHahm commented on core/include/thread_flags.h:
```diff
@@ -0,0 +1,145 @@
+ *
+ * Note that some flags (currently the three most significant bits) are used by
+ * core functions and should not be set by the user. They can be waited for.
+ *
```
In general: nice documentation, thanks for the effort. Appreciated.

OlegHahm commented on core/include/thread_flags.h:
```diff
@@ -0,0 +1,145 @@
+thread_flags_t thread_flags_clear(thread_flags_t mask);
+
+/**
+ * @brief Wait for any flag in mask to become set (blocking)
```
Couldn't it make sense to provide also a non-blocking version?

E.g. I wait for event aka flag X and handle it. Now, before I go on, I want to make sure, that it didn't occur again -> `thread_flags_try_wait()` or similar?

kaspar030 replied:
Interesting.
I think ```thread_flags_clear()``` can be used for that, as it returns the flags it actually cleared. If that would suit your need, I'll add a note for that.

OlegHahm commented on core/include/thread_flags.h:
```diff
@@ -0,0 +1,145 @@
+thread_flags_t thread_flags_wait_any(thread_flags_t mask);
+
+/**
+ * @brief Wait for all flags in mask to become set (blocking)
```
What would be a potential use case for that?

OlegHahm replied:
Forget it, my fault.

kaspar030 replied:
Hm. You want to be notified as soon as there's incoming data from a tcp connection *and* the target file is writable?
You'd like to send a periodic data set, but only if there's a network connection?

OlegHahm replied:
Yes, sorry, I overread the "in mask" part and thought this function would wait for *all* flags to be set.

OlegHahm commented on core/include/thread_flags.h:
```diff
@@ -0,0 +1,145 @@
+ * Does not trigger yield.
+ *
+ * @param[in]   thread  thread to possibly wake up
+ * @return      1       if thread has been woken up
```
`s/thread/@p thread/`

OlegHahm commented on core/include/thread_flags.h:
```diff
@@ -0,0 +1,145 @@
+thread_flags_t thread_flags_wait_one(thread_flags_t mask);
+
+/**
+ * @brief Possibly wake up thread
```
lol, I know what this is supposed to say, but as is it reads just funny.

OlegHahm replied:
And I still think you should add an explanation here otherwise it's more like "this function may wake up the thread - or it may not, good luck!".

OlegHahm commented:
IMO you can squash. And if you confirm that the WIP label is not appropriate any more I would even ACK it.

OlegHahm commented on core/include/thread_flags.h:
```diff
@@ -0,0 +1,149 @@
+ * its flags field. Those flags can be set or unset, using thread_flags_set(),
+ * from ISR's, other threads or even by the thread itself.
+ *
+ * A thread can wait for any combination of it's flags to become set, using
```
here's another `it's` vs. `its` thing.

kaspar030 replied:
yup

OlegHahm commented on core/include/thread_flags.h:
```diff
@@ -0,0 +1,149 @@
+ * This API can be used to notify threads of conditions in a race-free
+ * and allocation-less way.
+ *
+ * Each thread has a number boolean flags (usually 16) in its tcb, stored in
```
what about this "usually 16"?

OlegHahm replied:
sorry - wasn't a good idea to ask you to squash

kaspar030 replied:
"usually as I thought about making thread_flags_t configurable, in case someone needs some more flags..."

Does it sound very confusing to you?

OlegHahm replied:
Well, I think it is a little bit confusing, because right now it is kind of fixed to 16 flags unless you change the typedef what I would not count as "configurable". See what I mean?

kaspar030 replied:
> sorry - wasn't a good idea to ask you to squash

no problemo, I like squashing. Prepare to get amended quickfixes, though. ;)

kaspar030 replied:
Wrote it and ammended the fixes to the wrong commit.

kaspar030 commented:
- addressed comments
- squashed

OlegHahm commented:
I'm fine with documentation now, too. (If we find an agreement about this `usually` thing.

kaspar030 commented:
> I'm fine with documentation now, too. (If we find an agreement about this usually thing.

I took out the "usually".

OlegHahm commented on core/include/thread_flags.h:
```diff
@@ -0,0 +1,149 @@
+ * This API can be used to notify threads of conditions in a race-free
+ * and allocation-less way.
+ *
+ * Each thread has a 16 boolean flags in its tcb, stored in its flags field as
```
"Each thread can handle up to 16 boolean flags, stored as a bitmask in the flags field of its tcb."

kaspar030 commented:
Nice late night reviewing, thanks!

OlegHahm commented:
I think it was about time that someone took a look and I can imagine that this feature could become quite handy in some situations.

kaspar030 commented:
- rebased

OlegHahm commented:
> core/include/thread_flags.h:107: trailing whitespace.

+ 

core/include/thread_flags.h:107: trailing whitespace.

+

kaspar030 commented:
- fixed and immediately squashed static-tests complaints

BytesGalore commented:
I guess the static tests fail because either of the undocumented flags [1] or the undocumented `typedef`s [1] in `thread_flags.h`.
For some reason T. seems to be a bit picky and very uninformative in such cases.

[1] https://github.com/RIOT-OS/RIOT/pull/4103/files#diff-d47cd507da62339e7c3e206a91cdc58aR65
[2] https://github.com/RIOT-OS/RIOT/pull/4103/files#diff-d47cd507da62339e7c3e206a91cdc58aR69

OlegHahm commented:
1. ACK

kaspar030 commented:
I thought nowadays one ack is enough? ;)

-----Original Message-----
From: Oleg Hahm <notifications@github.com>
To: RIOT-OS/RIOT <RIOT@noreply.github.com>
Cc: Kaspar Schleiser <kaspar@schleiser.de>
Sent: Di., 01 März 2016 21:23
Subject: Re: [RIOT] core: add thread flags (#4103)

1. ACK

---
Reply to this email directly or view it on GitHub:
https://github.com/RIOT-OS/RIOT/pull/4103#issuecomment-190886007

OlegHahm commented:
One ACK plus Travis is fine with me here.

gebart commented:
core/include/thread_flags.h:68: warning: Member tcb_t (typedef) of file thread_flags.h is not documented.

kaspar030 commented:
- rebased

Unfortunately now this increases the thread_t size again. I'm inclined to increase ```thread_flags_t``` tp 4 bytes again, as on 32bit platforms it doesn't use more space in thread_t.

DipSwitch commented:
Needs squash :)

kaspar030 commented:
- rebased, squashed

cgundogan commented:
needs a second rebase

kaspar030 commented:
(cancelling so we can see #5040)

haukepetersen commented:
I this still going into the release? Or should we postpone it?

kaspar030 commented:
> I this still going into the release? Or should we postpone it?

let me rebase and have another look. IMHO it should be ready.

kaspar030 commented:
- rebased on 4923

@haukepetersen IMHO ready to merge.

haukepetersen commented on core/include/thread_flags.h:
```diff
@@ -0,0 +1,155 @@
+thread_flags_t thread_flags_wait_all(thread_flags_t mask);
+
+/**
+ * @brief Wait for any flags in mask to become set (blocking), one at a time
```
most or least significant first? Please document

haukepetersen replied:
addressed

haukepetersen commented on tests/thread_flags/main.c:
```diff
@@ -0,0 +1,74 @@
+    _set(thread, 0x2);
+    _set(thread, 0x4);
+
+    while(1);
```
missed this one: please change to `while(1) {}`

haukepetersen commented:
Code looks good and works as expected (at least the test application)

ACK when addressed, rebased and squashed.

OlegHahm commented:
Needs a rebase after #4923 got merged.

kaspar030 commented:
- rebased

kaspar030 commented:
Dammit, removed the "waiting for pr" label too late...

kaspar030 commented:
@cgundogan would you take another look?

kaspar030 commented:
@cgundogan Ah, not needed, there are already two ACKs. ;)

cgundogan commented:
if you don't want my personal ACK (;

kaspar030 commented:
Don't want to overuse your ACKs, I know reviewing is hard! :)

cgundogan commented on core/thread_flags.c:
```diff
@@ -0,0 +1,135 @@
+{
+    unsigned state = irq_disable();
+    mask &= thread->flags;
+    thread->flags &= ~mask;
```
I am not 100% sure, but I think
```
mask &= thread->flags;
thread->flags &= ~mask;
```
can be reduced to `thread->flags &= ~mask;` only?

kaspar030 replied:
No, as the function needs to return the actually cleared flags.

cgundogan replied:
dammit, I didn't look that far..

kaspar030 commented:
Murdock is fine -> go.

cgundogan commented on core/thread_flags.c:
```diff
@@ -0,0 +1,135 @@
+    _thread_flags_wait_any(mask);
+    thread_t *me = (thread_t*) sched_active_thread;
+    unsigned tmp = me->flags & mask;
+    return _thread_flags_clear_atomic(me, thread_flags_clear(1 << bitarithm_lsb(tmp)));
```
If you are trying to unset all bits, but not the least significant one, then you may use `tmp &= -tmp`, or `tmp &= ~tmp + 1` (equivalent) instead of `1 << bitarithm_lsb(tmp)`.

kaspar030 replied:
Well, I'm trying to unset *only* the least signigicant bit. :)

cgundogan replied:
`1 << bitarithm_lsb(tmp)` isn't this the same as unsetting all bits but not the lsb?

kaspar030 replied:
Ah yes.

```
>>> a = 0x1110;  a &= -a; print(bin(a))
0b10000
>>>
```

Is this due to pythons signed bit handling?

cgundogan replied:
yes, this only works if the number is unsigned. The number gets negated and 1 is added to the result for  negative unsigned numbers .. AFAIK

kaspar030 replied:
See #5200. Saves a whopping 32 bytes of code. ;)

kaspar030 replied:
(16byte on 32bit ARM).
